stackslib/src/chainstate/stacks/boot/signers-voting.clar

```diff
@@ -62,32 +73,84 @@
+            )
+            (map-set cycle-total-weight reward-cycle total)
+            (ok total)
+        )
+    )
+)
+
+;; Signer vote for the aggregate public key of the next reward cycle
+;;  The vote happens in the prepare phase of the current reward cycle but may be ran more than
+;;  once resulting in different 'rounds.' Each signer vote is based on the weight of stacked
+;;  stx tokens & fetched from the .signers contract. The vote is ran until the consensus 
+;;  threshold of 70% for a specific aggregate public key is reached.
+(define-public (vote-for-aggregate-public-key (signer-index uint) (key (buff 33)) (round uint) (reward-cycle uint))
+    (let ((tally-key {reward-cycle: reward-cycle, round: round, aggregate-public-key: key})
+            ;; vote by signer weight
+            (signer-weight (try! (get-current-signer-weight signer-index)))
+            (new-total (+ signer-weight (default-to u0 (map-get? tally tally-key))))
+            (total-weight (try! (get-total-weight reward-cycle))))
+        ;; Check that key isn't already set
+        (asserts! (is-none (map-get? aggregate-public-keys reward-cycle)) (err ERR_OUT_OF_VOTING_WINDOW))
+        ;; Check that the aggregate public key is correct length
+        (asserts! (is-eq (len key) u33) (err ERR_ILL_FORMED_AGGREGATE_PUBLIC_KEY))
+        ;; Check that aggregate public key has not been used before
+        (asserts! (is-valid-aggregate-public-key key {reward-cycle: reward-cycle, round: round}) (err ERR_DUPLICATE_AGGREGATE_PUBLIC_KEY))
+        ;; Check that signer hasn't voted in reward-cycle & round
+        (asserts! (map-insert votes {reward-cycle: reward-cycle, round: round, signer: tx-sender} {aggregate-public-key: key, signer-weight: signer-weight}) (err ERR_DUPLICATE_VOTE))
+        ;; Update tally aggregate public key candidate
         (map-set tally tally-key new-total)
+        ;; Update used aggregate public keys
         (map-set used-aggregate-public-keys key {reward-cycle: reward-cycle, round: round})
         (update-last-round reward-cycle round)
-        (print { 
-            event: "voted", 
-            signer: tx-sender, 
-            reward-cycle: reward-cycle, 
-            round: round, 
-            key: key, 
-            new-total: new-total })
+        (print {
+            event: "voted",
+            signer: tx-sender,
+            reward-cycle: reward-cycle,
+            round: round,
+            key: key,
+            new-total: new-total,
+        })
+        ;; Check if consensus has been reached
+        (and
+            ;; If we already have consensus, skip this
+            (is-none (map-get? aggregate-public-keys reward-cycle))
+            ;; If the new total weight is greater than or equal to the threshold consensus
+            (>= (/ (* new-total u1000) total-weight) threshold-consensus)
+            ;; Save this approved aggregate public key for this reward cycle
+            (map-set aggregate-public-keys reward-cycle key)
```

Comment:
Why not use a "map-insert" here since it can only be set once & not overwritten?

Reply:
Good point, we could just remove the `(is-none (map-get? ...))` and just use the `map-insert` to break the `and` if it is already set.

Reply:
Changed in d8c8cfafd09cc3864eb2a61cd6f22f28bf3a51ce.